dkc/core/rest/file.py

```diff
@@ -91,4 +93,6 @@ def perform_create(self, serializer: FileSerializer):
     @action(detail=True)
     def download(self, request, pk=None):
         file = get_object_or_404(File, pk=pk)
-        return HttpResponseRedirect(file.blob.url)
+        if file.blob:  # FileFields are falsy when not populated with a file
+            return HttpResponseRedirect(file.blob.url)
+        return FileResponse(BytesIO(), filename=file.name, content_type=file.content_type)
```

Comment:
If the file hasn't been uploaded yet, I think this should return a 202, 204, or 404 response. To me, a partially uploaded file shouldn't be considered a fully valid resource by end API users.

Reply:
Good point, I'm not sure why I didn't think of something like 204. I wanted to specifically avoid 4xx, thinking of someone trying to sync a tree with a partially uploaded file, and I didn't want to break the sync operation if a pending file exists.

Reply:
I don't think I've fully thought this through, but don't we want file upload to appear to be atomic from the perspective of read-only API endpoints? That is, the file will be non-existent / invisible until the blob (and potentially the hashsum?) is ready.

Reply:
Perhaps so, but that's a lot more complicated to engineer. If you have a design in mind, please detail it.

Reply:
As a start, we could filter the `FileViewSet` queryset for listing / details to always exclude anything without a `blob`; that would cause partial files to effectively be invisible in a folder listing. However the folder itself would already have its size incremented, so an attentive observer would notice that certain folder sizes were listed as larger as the sum of their contents.

From a user experience perspective, I'm not sure which is preferable.

I agree that to have fully robust atomicity, we'd have to maintain a separate set of quota reservations or have more complicated logic for reporting effective folder sizes.